client/components/data/query-gsuite-users/index.jsx

```diff
@@ -2,34 +2,36 @@
  * External dependencies
  */
 import PropTypes from 'prop-types';
-import { useEffect } from 'react';
-import { connect } from 'react-redux';
+import { useEffect, useRef } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
 
 /**
  * Internal dependencies
  */
 import { getGSuiteUsers } from 'state/gsuite-users/actions';
 import isRequestingGSuiteUsers from 'state/selectors/is-requesting-gsuite-users';
 
-const QueryGSuiteUsers = ( { siteId, request, isRequesting } ) => {
+const QueryGSuiteUsers = ( { siteId } ) => {
+	const dispatch = useDispatch();
+	const isRequesting = useSelector( state => isRequestingGSuiteUsers( state, siteId ) );
+	const prevIsRequesting = useRef( isRequesting );
+
 	useEffect( () => {
-		if ( ! isRequesting ) {
-			request( siteId );
+		if ( ! ( isRequesting || prevIsRequesting.current ) ) {
+			dispatch( getGSuiteUsers( siteId ) );
```

Comment:
Checking also the ref value will ensure that the request won't be issued if the `isRequesting` flag has ever been `true` during the lifetime of the `Query` component? In other words, the request is sent at most once during the lifetime of the component, and only if on initial mount the `isRequesting` value was `false`?

That fails to take into account a change of `siteId`. If the `Query` component is rerended with a different `siteId`, it should issue the REST request for the new `siteId`.

With @sgomes, we worked on several `Query` component updates recently, in #37521. And the best practice that emerged is the following pattern:

1) Use effect and Redux hooks to issue a request on mount and when a relevant prop changes:
```js
function QueryGSuiteUsers( { siteId } ) {
	const dispatch = useDispatch();

	useEffect( () => {
		dispatch( request( siteId ) );
	}, [ siteId, dispatch ] );

	return null;
}
```

2) Prevent issuing a second request while the first one is still in flight. Using Redux action thunk:
```js
const request = siteId => ( dispatch, getState ) => {
	const isRequesting = isRequestingGSuiteUsers( getState(), siteId );
	if ( ! isRequesting ) {
		dispatch( getGSuiteUsers( siteId ) );
	}
};
```

The two above techniques (hooks + thunk) should work perfectly.

The original `QueryGSuiteUsers` had a serious and obvious bug: the dependencies of the effect included `isRequesting`:
```js
useEffect( () => {
  if ( ! isRequesting ) request( siteId );
}, [ siteId, request, isRequesting ] );
```
That means that the effect is run on any change of `isRequesting`, including a transition from `true` to `false` when the request is finished! That immediately starts the request again.

The thunk technique solves that: note that `isRequesting` is no longer a dependency of the hook -- it's not used anywhere in the React component. Instead, it's used in the thunk, at the Redux level.

Reply:
Thank you very much for providing such a detailed feedback Jarda! We will update the component per your recommendation 👍.

Reply:
Likewise, thanks Jarda!